test/mpi/test_mpi_cuda.c

```
@@ -0,0 +1,668 @@
+        } \
+    } while (0)
+
+typedef struct {
+    CUdeviceptr ptr;
+    size_t      size;
+    void        *obj;
+} alloc_mem_t;
+
+typedef struct {
+    const char *name;
+    alloc_mem_t (*alloc)(size_t);
+    void (*free)(alloc_mem_t*);
+} allocator_t;
+
+typedef struct {
+    const char     *description;
+    const unsigned min_gpus;
+    const int      mthread_support;
+    int (*func)(const allocator_t*, const allocator_t*, const void*, size_t,
+                CUdevice);
+} test_t;
+
+typedef struct {
+    CUdeviceptr d_send;
+    CUdeviceptr d_recv;
+    size_t      size;
+} thread_arg_t;
+
+const size_t sizes[] = {SIZE_S, SIZE_M, SIZE_L};
+
+static unsigned total_ucx_errors_and_warnings = 0;
+
+static void *create_gold_data(size_t size)
+{
+    uint8_t *gold_data;
+    int i;
+
+    gold_data = malloc(size);
+    if (gold_data == NULL) {
+        fprintf(stderr, "failed to create gold data\n");
+        exit(-1);
+    }
+
+    for (i = 0; i < size; ++i) {
+        gold_data[i] = i % 256;
+    }
+
+    return gold_data;
+}
+
+static ucs_log_func_rc_t
+ucx_errors_and_warnings_counter(const char *file, unsigned line,
+                                const char *function, ucs_log_level_t level,
+                                const ucs_log_component_config_t *comp_conf,
+                                const char *format, va_list ap)
+{
+    if ((level == UCS_LOG_LEVEL_ERROR) || (level == UCS_LOG_LEVEL_WARN)) {
+        ++total_ucx_errors_and_warnings;
+    }
+
+    return UCS_LOG_FUNC_RC_CONTINUE;
+}
+
+static void
+mpi_send_recv(int sender, CUdeviceptr d_send, CUdeviceptr d_recv, size_t size)
+{
+    int rank, receiver;
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    receiver = (sender + 1) % 2;
+
+    if (rank == sender) {
+        MPI_Send((void*)d_send, size, MPI_BYTE, receiver, 0, MPI_COMM_WORLD);
+    } else {
+        MPI_Recv((void*)d_recv, size, MPI_BYTE, sender, 0, MPI_COMM_WORLD,
+                 MPI_STATUS_IGNORE);
```

maybe it is worth checking return values here?

---

to make sure we do not miss an error if it silently fails